File: sdk/typescript/src/query.ts
/**
 * Claude Agent SDK–compatible `query()` entrypoint.
 *
 * ```ts
 * import { query } from "@recursive/sdk";
 *
 * for await (const message of query({
 *   prompt: "Find and fix the bug in auth.ts",
 *   options: { maxTurns: 10, permissionMode: "bypassPermissions" },
 * })) {
 *   if (message.type === "assistant") { ... }
 *   if (message.type === "result") { console.log(message.result); }
 * }
 * ```
 *
 * Spawns `recursive -p … --output-format stream-json --input-format stream-json`
 * (no `-H`) so the bidirectional control channel is open for `canUseTool`,
 * hooks, `interrupt`, and streaming-input follow-up turns.
 */

import {
  spawnControlSession,
  type CanUseTool,
  type ControlSessionHandle,
  type ControlSpawnOptions,
  type HookCallback,
  type PermissionResult,
} from "./controlSession.js";
import type { WireItem } from "./wire.js";

// ── Options (Claude Agent SDK–aligned names) ──────────────────────────────

/**
 * Subset of Claude Agent SDK `Options` that Recursive can honour today.
 */
export interface Options {
  /** Working directory for the agent (`--workspace`). */
  cwd?: string;
  /** Model id (`-m`). */
  model?: string;
  /** Max agent steps (`--max-steps`). Claude name: `maxTurns`. */
  maxTurns?: number;
  /**
   * System prompt. String replaces the default; preset object with `append`
   * appends to Recursive's built-in prompt.
   */
  systemPrompt?:
    | string
    | { type: "preset"; preset: "claude_code" | string; append?: string };
  /**
   * Permission mode. Claude values map onto Recursive CLI flags:
   * - `bypassPermissions` / `acceptEdits` / `dontAsk` / `auto` → `auto`
   * - `plan` → plan-first mode
   * - `default` → `default`
   */
  permissionMode?:
    | "default"
    | "acceptEdits"
    | "bypassPermissions"
    | "plan"
    | "dontAsk"
    | "auto"
    | "strict"
    | "bypass";
  /** Resume an existing session id (`-r`). */
  resume?: string;
  /** Path to the `recursive` binary (Claude: `pathToClaudeCodeExecutable`). */
  pathToClaudeCodeExecutable?: string;
  /** Alias for {@link Options.pathToClaudeCodeExecutable}. */
  executable?: string;
  /** AbortController to cancel the run. */
  abortController?: AbortController;
  /** Max API spend in USD (`--max-budget-usd`). */
  maxBudgetUsd?: number;
  /** Restrict tools (`--allow-tools`). */
  allowedTools?: string[];
  /**
   * Host-side tool permission callback (Claude `canUseTool`).
   * Required for interactive prompts; when omitted, tools are auto-allowed
   * under `bypassPermissions`/`auto` (or when listed in `allowedTools`).
   */
  canUseTool?: CanUseTool;
  /**
   * SDK hooks. Recursive forwards `hook_callback` control requests to the
   * registered callbacks after sending `initialize` with callback ids.
   */
  hooks?: {
    [event: string]: Array<{
      matcher?: string;
      hooks: HookCallback[];
    }>;
  };
}

export type { CanUseTool, PermissionResult, HookCallback };

// ── Message types yielded by query() ──────────────────────────────────────

/**
 * Claude-compatible stream message.
 *
 * Includes `type: "result"` as a stream item (Claude Agent SDK behaviour),
 * not a separate `wait()` return value.
 */
export type QueryMessage = Record<string, unknown> & {
  type: string;
  session_id?: string;
};

// ── Query ─────────────────────────────────────────────────────────────────

/**
 * Return type of {@link query}. An async generator of {@link QueryMessage}
 * plus control methods.
 */
export interface Query extends AsyncGenerator<QueryMessage, void> {
  /** Interrupt the running agent (control `interrupt` + SIGTERM). */
  interrupt(): Promise<void>;
  /** Close stdin / tear down the underlying process. */
  close(): void;
  /** Stream additional user turns into the same CLI session. */
  streamInput(prompts: AsyncIterable<string>): Promise<void>;
  /** Host→CLI `set_permission_mode`. */
  setPermissionMode(mode: string): Promise<void>;
  /** Host→CLI `set_model`. */
  setModel(model: string): Promise<void>;
}

/**
 * Run an agent turn and stream Claude-compatible messages.
 *
 * Recursive equivalent of `@anthropic-ai/claude-agent-sdk`'s `query()`.
 */
export function query(params: {
  prompt: string | AsyncIterable<string>;
  options?: Options;
}): Query {
  const options = params.options ?? {};
  const { hookCallbacks, initializeHooks } = materializeHooks(options.hooks);

  const isStreamPrompt =
    typeof params.prompt !== "string" &&
    params.prompt != null &&
    typeof (params.prompt as AsyncIterable<string>)[Symbol.asyncIterator] ===
      "function";

  const initialPrompt = typeof params.prompt === "string" ? params.prompt : "";
  const spawnOpts = optionsToControlSpawn(initialPrompt || " ", options, {
    hookCallbacks,
    initializeHooks,
    keepStdinOpen: isStreamPrompt,
  });

  const handle = spawnControlSession(spawnOpts);

  if (options.abortController) {
    const { signal } = options.abortController;
    if (signal.aborted) {
      handle.cancel();
    } else {
      signal.addEventListener("abort", () => void handle.interrupt(), {
        once: true,
      });
    }
  }

  let streamInputDone: Promise<void> | undefined;
  if (isStreamPrompt) {
    streamInputDone = (async () => {
      for await (const text of params.prompt as AsyncIterable<string>) {
        handle.writeUser(text);
      }
      handle.close();
    })();
  }

  const gen = (async function* (): AsyncGenerator<QueryMessage, void> {
    try {
      for await (const item of handle.items()) {
        const msg = wireItemToQueryMessage(item);
        if (msg) yield msg;
      }
    } finally {
      if (streamInputDone) await streamInputDone.catch(() => undefined);
    }
  })();

  return Object.assign(gen, {
    async interrupt(): Promise<void> {
      await handle.interrupt();
    },
    close(): void {
      handle.close();
      handle.cancel();
    },
    async streamInput(prompts: AsyncIterable<string>): Promise<void> {
      for await (const text of prompts) {
        handle.writeUser(text);
      }
      handle.close();
    },
    async setPermissionMode(mode: string): Promise<void> {
      await handle.setPermissionMode(mode);
    },
    async setModel(model: string): Promise<void> {
      await handle.setModel(model);
    },
  }) as Query;
}

// ── helpers ───────────────────────────────────────────────────────────────

/** @internal — maps Claude options onto control-session spawn opts. */
export function optionsToSpawn(
  prompt: string,
  options: Options,
): ControlSpawnOptions {
  return optionsToControlSpawn(prompt, options, {});
}

function optionsToControlSpawn(
  prompt: string,
  options: Options,
  extras: {
    hookCallbacks?: Map<string, HookCallback>;
    initializeHooks?: Record<string, Array<{ hookCallbackIds: string[] }>>;
    keepStdinOpen?: boolean;
  },
): ControlSpawnOptions {
  let systemPrompt: string | undefined;
  let appendSystemPrompt: string | undefined;
  if (typeof options.systemPrompt === "string") {
    systemPrompt = options.systemPrompt;
  } else if (options.systemPrompt && typeof options.systemPrompt === "object") {
    appendSystemPrompt = options.systemPrompt.append;
  }

  const spawn: ControlSpawnOptions = {
    prompt,
    cwd: options.cwd,
    model: options.model,
    maxSteps: options.maxTurns,
    maxBudgetUsd: options.maxBudgetUsd,
    systemPrompt,
    appendSystemPrompt,
    resumeSessionId: options.resume,
    cliPath: options.pathToClaudeCodeExecutable ?? options.executable,
    permissionMode: mapClaudePermission(options.permissionMode),
    allowedTools: options.allowedTools,
    canUseTool: options.canUseTool,
    hookCallbacks: extras.hookCallbacks,
    initializeHooks: extras.initializeHooks,
    abortSignal: options.abortController?.signal,
    keepStdinOpen: extras.keepStdinOpen,
  };

  if (options.permissionMode === "plan") {
    spawn.planningMode = "plan_first";
    spawn.permissionMode = "default";
  }

  return spawn;
}

function mapClaudePermission(
  mode: Options["permissionMode"],
): ControlSpawnOptions["permissionMode"] {
  switch (mode) {
    case "bypassPermissions":
    case "acceptEdits":
    case "dontAsk":
    case "auto":
    case "bypass":
      return "auto";
    case "plan":
      return "default";
    case "strict":
      return "strict";
    case "default":
    case undefined:
      return "default";
    default:
      return "default";
  }
}

function materializeHooks(hooks: Options["hooks"]): {
  hookCallbacks: Map<string, HookCallback>;
  initializeHooks: Record<string, Array<{ hookCallbackIds: string[] }>>;
} {
  const hookCallbacks = new Map<string, HookCallback>();
  const initializeHooks: Record<
    string,
    Array<{ hookCallbackIds: string[] }>
  > = {};
  if (!hooks) return { hookCallbacks, initializeHooks };

  let n = 0;
  for (const [event, matchers] of Object.entries(hooks)) {
    const ids: string[] = [];
    for (const matcher of matchers) {
      for (const cb of matcher.hooks) {
        const id = `hook_${event}_${n++}`;
        hookCallbacks.set(id, cb);
        ids.push(id);
      }
    }
    if (ids.length > 0) {
      initializeHooks[event] = [{ hookCallbackIds: ids }];
    }
  }
  return { hookCallbacks, initializeHooks };
}

function wireItemToQueryMessage(item: WireItem): QueryMessage | null {
  if (item.kind === "result") {
    const r = item.result;
    const base: QueryMessage = {
      type: "result",
      subtype: r.subtype,
      is_error: !r.ok,
      session_id: r.id,
      num_turns: r.numTurns,
      duration_ms: r.durationMs,
      stop_reason: r.finishReason,
    };
    if (r.result !== undefined) base["result"] = r.result;
    if (r.usage) {
      base["usage"] = {
        input_tokens: r.usage.inputTokens,
        output_tokens: r.usage.outputTokens,
        cache_read_input_tokens: r.usage.cacheReadTokens,
        cache_creation_input_tokens: r.usage.cacheCreationTokens,
      };
    }
    if (r.error) base["errors"] = [r.error];
    return base;
  }

  if (item.kind !== "message") return null;

  const m = item.message;
  if (m.type === "assistant") {
    return {
      type: "assistant",
      session_id: m.sessionId,
      parent_tool_use_id: null,
      message: { role: "assistant", content: m.content },
    };
  }
  if (m.type === "user") {
    return {
      type: "user",
      session_id: m.sessionId,
      parent_tool_use_id: null,
      message: { role: "user", content: m.content },
    };
  }
  if (m.type === "system") {
    return {
      type: "system",
      subtype: m.subtype,
      session_id:
        typeof m.data["session_id"] === "string"
          ? String(m.data["session_id"])
          : undefined,
      ...m.data,
    };
  }
  if (m.type === "stream_event") {
    return {
      type: "stream_event",
      session_id: m.sessionId,
      parent_tool_use_id: null,
      event: {
        type: "content_block_delta",
        index: m.step,
        delta: { type: "text_delta", text: m.text },
      },
    };
  }
  if (m.type === "tool_progress") {
    return {
      type: "tool_progress",
      session_id: m.sessionId,
      tool_use_id: m.toolUseId,
      tool_name: m.toolName,
      elapsed_time_seconds: m.elapsedMs / 1000,
    };
  }
  return null;
}

/** @internal — exported for tests */
export type { ControlSessionHandle };

File: sdk/python/recursive_sdk/query.py
"""Claude Agent SDK–compatible ``query()`` entrypoint.

Example::

    import asyncio
    from recursive_sdk import query, ClaudeAgentOptions

    async def main():
        async for message in query(
            prompt="Find and fix the bug in auth.py",
            options=ClaudeAgentOptions(
                max_turns=10,
                permission_mode="bypassPermissions",
            ),
        ):
            if message.get("type") == "result":
                print(message.get("result"))

    asyncio.run(main())
"""

from __future__ import annotations

import asyncio
import threading
from dataclasses import dataclass, field
from typing import Any, AsyncIterator, Callable, Dict, List, Optional, Union

from .control_session import spawn_control_session

CanUseTool = Callable[..., Any]
HookCallback = Callable[..., Any]


@dataclass
class ClaudeAgentOptions:
    """Subset of Claude Agent SDK options that Recursive can honour today."""

    cwd: Optional[str] = None
    model: Optional[str] = None
    max_turns: Optional[int] = None
    system_prompt: Optional[Union[str, Dict[str, Any]]] = None
    permission_mode: Optional[str] = None
    resume: Optional[str] = None
    path_to_cli: Optional[str] = None
    """Path to the ``recursive`` binary (Claude: path_to_claude_code_executable)."""
    max_budget_usd: Optional[float] = None
    allowed_tools: List[str] = field(default_factory=list)
    can_use_tool: Optional[CanUseTool] = None
    hooks: Optional[Dict[str, Any]] = None
    # Accepted for call-site portability; mid-run MCP apply is CLI best-effort.
    mcp_servers: Dict[str, Any] = field(default_factory=dict)


class Query:
    """Async iterator with Claude-style control methods (sync constructor)."""

    def __init__(self, handle: Any) -> None:
        self._handle = handle
        self._queue: Optional[asyncio.Queue[Optional[Dict[str, Any]]]] = None
        self._thread: Optional[threading.Thread] = None
        self._started = False

    def _ensure_started(self) -> asyncio.Queue[Optional[Dict[str, Any]]]:
        if self._started and self._queue is not None:
            return self._queue
        loop = asyncio.get_running_loop()
        queue: asyncio.Queue[Optional[Dict[str, Any]]] = asyncio.Queue()
        handle = self._handle

        def _worker() -> None:
            try:
                for item in handle.items():
                    msg = _wire_item_to_query_message(item)
                    if msg is not None:
                        loop.call_soon_threadsafe(queue.put_nowait, msg)
            except Exception as err:  # noqa: BLE001
                loop.call_soon_threadsafe(
                    queue.put_nowait,
                    {
                        "type": "result",
                        "subtype": "error_during_execution",
                        "is_error": True,
                        "errors": [str(err)],
                    },
                )
            finally:
                loop.call_soon_threadsafe(queue.put_nowait, None)

        self._thread = threading.Thread(target=_worker, daemon=True)
        self._thread.start()
        self._queue = queue
        self._started = True
        return queue

    def __aiter__(self) -> "Query":
        return self

    async def __anext__(self) -> Dict[str, Any]:
        queue = self._ensure_started()
        try:
            msg = await queue.get()
        except (asyncio.CancelledError, GeneratorExit):
            self._handle.cancel()
            raise
        if msg is None:
            if self._thread is not None:
                self._thread.join(timeout=5.0)
            raise StopAsyncIteration
        return msg

    async def interrupt(self) -> None:
        await asyncio.to_thread(self._handle.interrupt)

    def close(self) -> None:
        self._handle.close()
        self._handle.cancel()

    async def stream_input(self, prompts: AsyncIterator[str]) -> None:
        async for text in prompts:
            await asyncio.to_thread(self._handle.write_user, text)
        await asyncio.to_thread(self._handle.close)

    async def set_permission_mode(self, mode: str) -> None:
        await asyncio.to_thread(self._handle.set_permission_mode, mode)

    async def set_model(self, model: str) -> None:
        await asyncio.to_thread(self._handle.set_model, model)


def query(
    *,
    prompt: str,
    options: Optional[ClaudeAgentOptions] = None,
) -> Query:
    """
    Run an agent turn and yield Claude-compatible message dicts.

    The terminal ``type: "result"`` object is yielded **inside** the stream
    (same contract as ``claude_agent_sdk.query``).
    """
    opts = options or ClaudeAgentOptions()
    hook_callbacks, initialize_hooks = _materialize_hooks(opts.hooks)
    spawn_kwargs = _options_to_spawn(prompt, opts)
    spawn_kwargs["hook_callbacks"] = hook_callbacks
    spawn_kwargs["initialize_hooks"] = initialize_hooks or None
    handle = spawn_control_session(**spawn_kwargs)
    return Query(handle)


def _materialize_hooks(
    hooks: Optional[Dict[str, Any]],
) -> tuple[Dict[str, HookCallback], Dict[str, Any]]:
    callbacks: Dict[str, HookCallback] = {}
    initialize: Dict[str, Any] = {}
    if not hooks:
        return callbacks, initialize
    n = 0
    for event, matchers in hooks.items():
        ids: List[str] = []
        if not isinstance(matchers, list):
            continue
        for matcher in matchers:
            if not isinstance(matcher, dict):
                continue
            for cb in matcher.get("hooks") or []:
                cb_id = f"hook_{event}_{n}"
                n += 1
                callbacks[cb_id] = cb
                ids.append(cb_id)
        if ids:
            initialize[event] = [{"hookCallbackIds": ids}]
    return callbacks, initialize


def _options_to_spawn(prompt: str, opts: ClaudeAgentOptions) -> Dict[str, Any]:
    system_prompt: Optional[str] = None
    append_system_prompt: Optional[str] = None
    if isinstance(opts.system_prompt, str):
        system_prompt = opts.system_prompt
    elif isinstance(opts.system_prompt, dict):
        append_system_prompt = opts.system_prompt.get("append")

    permission_mode = _map_permission(opts.permission_mode)
    planning_mode = None
    if opts.permission_mode == "plan":
        planning_mode = "plan_first"
        permission_mode = "default"

    return {
        "prompt": prompt,
        "cwd": opts.cwd,
        "model": opts.model,
        "max_steps": opts.max_turns,
        "max_budget_usd": opts.max_budget_usd,
        "system_prompt": system_prompt,
        "append_system_prompt": append_system_prompt,
        "resume_session_id": opts.resume,
        "cli_path": opts.path_to_cli,
        "permission_mode": permission_mode,
        "planning_mode": planning_mode,
        "allowed_tools": opts.allowed_tools or None,
        "can_use_tool": opts.can_use_tool,
    }


def _map_permission(mode: Optional[str]) -> Optional[str]:
    if mode in (
        "bypassPermissions",
        "acceptEdits",
        "dontAsk",
        "auto",
        "bypass",
    ):
        return "auto"
    if mode == "strict":
        return "strict"
    if mode in ("default", "plan", None):
        return "default"
    return "default"


def _wire_item_to_query_message(item: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    kind = item.get("kind")
    if kind == "result":
        r = item["result"]
        msg: Dict[str, Any] = {
            "type": "result",
            "subtype": r.subtype,
            "is_error": not r.ok,
            "session_id": r.id,
            "num_turns": r.num_turns,
            "duration_ms": r.duration_ms,
            "stop_reason": r.finish_reason,
        }
        if r.result is not None:
            msg["result"] = r.result
        if r.usage is not None:
            msg["usage"] = {
                "input_tokens": r.usage.input_tokens,
                "output_tokens": r.usage.output_tokens,
                "cache_read_input_tokens": r.usage.cache_read_tokens,
                "cache_creation_input_tokens": r.usage.cache_creation_tokens,
            }
        if r.error:
            msg["errors"] = [r.error]
        return msg

    if kind != "message":
        return None

    m = item["message"]
    mtype = getattr(m, "type", None)
    if mtype == "assistant":
        content = []
        for b in m.content:
            if getattr(b, "type", None) == "text":
                content.append({"type": "text", "text": b.text})
            else:
                content.append(
                    {
                        "type": "tool_use",
                        "id": getattr(b, "id", ""),
                        "name": getattr(b, "name", ""),
                        "input": getattr(b, "input", {}),
                    }
                )
        return {
            "type": "assistant",
            "session_id": m.session_id,
            "parent_tool_use_id": None,
            "message": {"role": "assistant", "content": content},
        }
    if mtype == "user":
        return {
            "type": "user",
            "session_id": m.session_id,
            "parent_tool_use_id": None,
            "message": {"role": "user", "content": m.content},
        }
    if mtype == "system":
        out = {"type": "system", "subtype": m.subtype, **m.data}
        if "session_id" in m.data:
            out["session_id"] = m.data["session_id"]
        return out
    if mtype == "stream_event":
        return {
            "type": "stream_event",
            "session_id": m.session_id,
            "parent_tool_use_id": None,
            "event": {
                "type": "content_block_delta",
                "index": m.step,
                "delta": {"type": "text_delta", "text": m.text},
            },
        }
    if mtype == "tool_progress":
        return {
            "type": "tool_progress",
            "session_id": m.session_id,
            "tool_use_id": m.tool_use_id,
            "tool_name": m.tool_name,
            "elapsed_time_seconds": m.elapsed_ms / 1000.0,
        }
    return None


# Exported for unit tests
options_to_spawn = _options_to_spawn
map_permission = _map_permission
wire_item_to_query_message = _wire_item_to_query_message
